ding/policy/mbpolicy/mbsac.py

```diff
@@ -219,3 +221,197 @@ def _monitor_vars_learn(self) -> List[str]:
+           ID Symbol                  Type        Default Value  Description
+           == ====================    ========    =============  =====================================
+           1  ``learn.grad_clip`      float       100.0          | Max norm of gradients.
+           2  ``learn.ensemble_size`` int         1              | The number of ensemble world models.
+           == ====================    ========    =============  =====================================
+
+        .. note::
+            For other configs, please refer to ding.policy.sac.SACPolicy.
+       """
+
+    config = dict(
+        learn=dict(
+            # (float) Max norm of gradients.
+            grad_clip=100,
+            # (int) The number of ensemble world models.
+            ensemble_size=1,
+        )
+    )
+
+    @classmethod
+    def default_config(cls: type) -> EasyDict:
+        cfg = copy.deepcopy(cls.config)
+        cfg = EasyDict(deep_merge_dicts(super().config, cfg))
+        cfg.cfg_type = cls.__name__ + 'Dict'
+        return cfg
+
+    def _init_learn(self) -> None:
+        super()._init_learn()
+
+        self._grad_clip = self._cfg.learn.grad_clip
+        self._ensemble_size = self._cfg.learn.ensemble_size
+
+        self._target_model.requires_grad_(False)
+
+        # TODO: auto alpha
+        self._auto_alpha = False
+
+        def actor_fn(obs: Tensor):
+            obs, dim = fold_batch(obs, 1)
+            (mu, sigma) = self._learn_model.forward(obs, mode='compute_actor')['logit']
+            dist = Independent(Normal(mu, sigma), 1)
+            pred = dist.rsample()
+            action = torch.tanh(pred)
+
+            log_prob = dist.log_prob(
+                pred
+            ) + 2 * (pred + torch.nn.functional.softplus(-2. * pred) - torch.log(torch.tensor(2.))).sum(-1)
+            aug_reward = -self._alpha.detach() * log_prob
+
+            return unfold_batch(action, dim), unfold_batch(aug_reward, dim)
+
+        self._actor_fn = actor_fn
+
+        def critic_fn(obss: Tensor, actions: Tensor, model: nn.Module):
+            eval_data = {'obs': obss, 'action': actions}
+            q_values = model.forward(eval_data, mode='compute_critic')['q_value']
+            return q_values
+
+        self._critic_fn = critic_fn
+
+    def _forward_learn(self, data: dict, world_model, envstep) -> Dict[str, Any]:
+        # preprocess data
+        data = default_preprocess_learn(
+            data,
+            use_priority=self._priority,
+            use_priority_IS_weight=self._cfg.priority_IS_weight,
+            ignore_done=self._cfg.learn.ignore_done,
+            use_nstep=False
+        )
+        if self._cuda:
+            data = to_device(data, self._device)
+
+        if len(data['action'].shape) == 1:
+            data['action'] = data['action'].unsqueeze(1)
+
+        # [B, D] -> [E, B, D]
+        data['next_obs'] = data['next_obs'].unsqueeze(0).repeat(self._ensemble_size, 1, 1)
```

Comment:
maybe a `repeat_batch` here

Reply:
Is `repeat_batch` a `pytorch` funtion?

Reply:
no, you need to create a function like `fold_batch`

Reply:
I create a helper function `unsqueeze_repeat`